### PR TITLE
Add pure-PyTorch reference implementations

### LINKED ISSUE #1177: Add a PyTorch reference implementation
To compare with current impl.

### PR DESCRIPTION
**Pull request by ryankert01**
Closes #1177 
Related to #1227 

## Summary

This PR adds a pure-PyTorch reference backend to the QDP Python package to compare with our implementation of GPU kernel~

## Benchmark Results

All runs: 100 batches x 64 vectors (except 18-qubit: 50 batches x 64), median of 3 trials.

### Amplitude Encoding

| Qubits | Mode | PyTorch CPU | PyTorch GPU | Mahout | Mahout vs GPU |
|---|---|---|---|---|---|
| 10 | encode-only | 462,882 | 1,390,998 | 567,499 | 0.4x |
| 10 | end-to-end | 73,699 | 86,939 | 234,170 | **2.7x** |
| 14 | encode-only | 118,620 | 789,862 | 151,713 | 0.2x |
| 14 | end-to-end | 5,514 | 6,356 | 25,603 | **4.0x** |
| 16 | encode-only | 5,458 | 228,525 | 65,358 | 0.3x |
| 16 | end-to-end | 721 | 964 | 6,336 | **6.6x** |
| 18 | encode-only | 1,313 | 58,761 | 15,876 | 0.3x |
| 18 | end-to-end | 194 | 237 | 1,529 | **6.5x** |

### Angle Encoding

| Qubits | Mode | PyTorch CPU | PyTorch GPU | Mahout | Mahout vs GPU |
|---|---|---|---|---|---|
| 14 | encode-only | 1,086 | 59,864 | 45,332 | 0.8x |
| 14 | end-to-end | 1,114 | 56,456 | 50,919 | 0.9x |
| 16 | encode-only | 262 | 10,254 | 8,032 | 0.8x |
| 16 | end-to-end | 252 | 10,093 | 11,496 | **1.1x** |

### IQP Encoding

| Qubits | Mode | PyTorch CPU | PyTorch GPU | Mahout | Mahout vs GPU |
|---|---|---|---|---|---|
| 10 | encode-only | 15,381 | 74,239 | 484,071 | **6.5x** |
| 14 | encode-only | 1,258 | 25,597 | 55,304 | **2.2x** |

### Analysis

- **Amplitude encode-only**: PyTorch GPU wins 2-4x because its vectorized L2-norm + pad is very efficient, while Mahout's `encode` path still pays per-batch GPU output allocation + D2H norm validation sync overhead.
- **Amplitude end-to-end**: Mahout wins 2.7-6.6x, with the advantage growing at higher qubit counts. Rust data generation + integrated pipeline dominates Python `generate_batch_data` + `torch.tensor` + H2D transfer.
- **Angle**: Near-parity in both modes (0.8-1.1x). The tensor-product encoding is compute-bound and both implementations are similarly efficient.
- **IQP encode-only**: Mahout wins decisively (2.2-6.5x). Mahout's CUDA kernel for IQP (Walsh-Hadamard + phase computation) is significantly faster than PyTorch's Python-level loop over butterfly stages.

### Known Limitations

- **Basis encode-only**: Rust `engine.encode` expects per-sample basis indices; batch input format differs from PyTorch. Requires Rust API change to support.
- **IQP end-to-end**: Rust pipeline uses `1 << num_qubits` as sample_size regardless of encoding method, causing a mismatch for IQP (which expects `n + n*(n-1)/2`). Pre-existing Rust pipeline bug.

### CODE REVIEW

**Comment by ryankert01** [comment]
cc @viiccwen @vvvdwbvvv for review

**Comment by ryankert01** [comment]
update at df90a6e

**Comment by ryankert01** [comment]
I already implements the amplitude one and successfully increase the speed to slightly faster than `torch.compile`.

File: qdp/qdp-python/qumat_qdp/torch_ref.py [hunk 1]
```diff
@@ -0,0 +1,359 @@
...
+    # Bit-pattern matrix: (state_dim, num_qubits)
+    # bits[i, k] = (i >> k) & 1
+    indices = torch.arange(state_dim, device=data.device, dtype=torch.long)
+    bits = (
+        (indices.unsqueeze(1) >> torch.arange(num_qubits, device=data.device)) & 1
+    ).to(data.dtype)
+
+    # For each state index: amplitude = prod_k (sin if bit else cos)
+    # Shape: (batch, state_dim, num_qubits) via broadcasting
+    trig = bits.unsqueeze(0) * sin_vals.unsqueeze(1) + (
```
Comment (rich7420) [inline qdp/qdp-python/qumat_qdp/torch_ref.py]: Nit: We can save GPU VRAM and skip arithmetic steps by using torch.where(bits.bool().unsqueeze(0), sin_vals.unsqueeze(1), cos_vals.unsqueeze(1)) instead of adding and multiplying here. maybe? I think

**Review by rich7420 — COMMENTED** [review]
Thanks for the patch! @ryankert01

File: qdp/qdp-python/qumat_qdp/torch_ref.py (hunk 1 above)
Comment (ryankert01) [reply qdp/qdp-python/qumat_qdp/torch_ref.py]: done

**Review by rich7420 — APPROVED** [review]
LGTM! thx for the update
